### PR TITLE
[Feat][Plugin] Enable Sparse MLA and GLM-5 for vLLM-ATOM

### PR DESCRIPTION
**Pull request by kliuae-amd**
## Motivation

Following #126, this PR enables sparse MLA in ATOM's vLLM plugin mode, adding support for GLM-5 models that uses index-based topk sparse attention.
## Technical Details

- Add Indexer and Sparse MLA backends for vLLM OOT plugin
- Register GLM-5 as supported models
## Test Plan

Accuracy test with lm_eval

Server command:
```
ATOM_DISABLE_VLLM_PLUGIN=0 \
ATOM_DISABLE_VLLM_PLUGIN_ATTENTION=0 \
VLLM_LOGGING_LEVEL=DEBUG \
VLLM_USE_V1=1 \
VLLM_ROCM_USE_AITER=1 \
vllm serve /path/to/GLM-5-FP8/ \
  -tp 8 \
  --max-num-seqs 1024 \
  --gpu-memory-utilization 0.9 \
  --no-enable-prefix-caching \
  --disable-uvicorn-access-log \
  --trust-remote-code \
  --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
  --kv-cache-dtype {auto,fp8} \
  --block-size 1
```
## Test Result

lm_eval command
```
lm_eval --model local-completions   --model_args model=/path/to/GLM-5-FP8/,base_url=http://localhost:8000/v1/completions   --batch_size 100  --tasks gsm8k --num_fewshot 20
```

Model: zai-org/GLM-5-FP8

ATOM
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|_  |0.9538|_  |0.0058|
|     |       |strict-match    |     5|exact_match|_  |0.9515|_  |0.0059|
```

vLLM Plugin (bf16 kv cache)
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|_  |0.953|_  |0.0058|
|     |       |strict-match    |    20|exact_match|_  |0.953|_  |0.0058|
```

vLLM Plugin (fp8 kv cache)
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|    20|exact_match|_  |0.9439|_  |0.0063|
|     |       |strict-match    |    20|exact_match|_  |0.9439|_  |0.0063|
```

Performance on MI300X, TP8
ISL/OSL | Concurrency | KV Cache | vLLM Plugin Req/s | ATOM Req/s | vLLM Plugin over ATOM (Req/s) | vLLM Plugin Total tok/s | ATOM Total tok/s | vLLM Plugin over ATOM (tok/s)
-- | -- | -- | -- | -- | -- | -- | -- | --
1k/1k | 128 | bf16 | 2.06 | 2.02 | +1.98% | 4224.31 | 4137.09 | +2.11%
1k/1k | 64 | bf16 | 1.40 | 1.36 | +2.94% | 2874.85 | 2784.97 | +3.23%
1k/1k | 128 | fp8 | 2.14 | 2.23 | -4.04% | 4383.43 | 4568.58 | -4.05%
1k/1k | 64 | fp8 | 1.44 | 1.43 | +0.70% | 2938.97 | 2935.14 | +0.13%

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: atom/plugin/vllm/model_wrapper.py [hunk 1]
```diff
@@ -104,6 +112,71 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
+           assign the allocated KV cache tensor
+        3. The indexer's metadata lookup uses the correct prefix in vLLM's
+           attn_metadata dict
+
+        ATOM's DeepseekV32IndexerCache inherits from nn.Module (not vLLM's
+        AttentionLayerBase), so we register it as a virtual subclass.
+        We also register each instance in vLLM's static_forward_context using
+        the same prefix convention as other attention layers (the prefix
+        parameter passed at construction, e.g. 'model.layers.0...k_cache').
+        """
```
Comment (wuhuikx) [inline atom/plugin/vllm/model_wrapper.py:149]: Thanks for the detailed comments here. I think it's very useful for developers to deep dive the code.

File: atom/plugin/vllm/register.py [hunk 2]
```diff
@@ -26,6 +26,7 @@
     "Glm4MoeForCausalLM": ATOM_MOE_CAUSAL_LM_MODEL_WRAPPER,
     "Qwen3_5ForConditionalGeneration": "atom.models.qwen3_5:Qwen3_5ForConditionalGeneration",
     "Qwen3_5MoeForConditionalGeneration": "atom.models.qwen3_5:Qwen3_5MoeForConditionalGeneration",
+    "GlmMoeDsaForCausalLM": ATOM_MOE_CAUSAL_LM_MODEL_WRAPPER,
```
Comment (wuhuikx) [inline atom/plugin/vllm/register.py]: Can we put the "GlmMoeDsaForCausalLM": ATOM_MOE_CAUSAL_LM_MODEL_WRAPPER, after "Glm4MoeForCausalLM": ATOM_MOE_CAUSAL_LM_MODEL_WRAPPER,

File: atom/model_ops/attention_mla.py [hunk 3]
```diff
@@ -203,6 +206,40 @@ def process_weights_after_loading(self):
             self.W_V, self.W_V_scale = dynamic_per_batched_tensor_quant(
                 W_V, dtype=dtypes.fp8
             )
+            
+            if is_plugin_mode() and is_vllm():
+                # The kernel operates on non-padded inputs. Hence, pre-compiling
+                # triton kernel to avoid runtime compilation for unseen batch sizes
+                # Pre-compile for batch sizes 1 to 1024 to cover most use-cases.
+                # On DS-R1, this step adds roughly 50s to the model loading time.
```
Comment (wuhuikx) [inline atom/model_ops/attention_mla.py]: I think it's a good idea here. how about the other triton kernel?

cc @valarLip @ZhangLirong-amd @ganyi1996ppo @zejunchen-zejun Can you help comment on this feature?

Comment (zejunchen-zejun) [reply atom/model_ops/attention_mla.py]: vLLM mainline precompiles the BMM kernel with different M before executing the model, I think we can leverage the Kuanfu's code here

File: atom/model_ops/attention_mla.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 import logging
 from dataclasses import dataclass
 from functools import partial as functools_partial
+from tqdm import tqdm
```
Comment (zejunchen-zejun) [inline atom/model_ops/attention_mla.py]: `tqdm` should be added into the ATOM requirement, otherwise when running ATOM native, there could be a dependency issue

File: atom/model_ops/attention_mla.py [hunk 5]
```diff
@@ -204,6 +209,41 @@ def process_weights_after_loading(self):
...
+                from vllm.distributed.parallel_state import is_global_first_rank
+
+                if is_global_first_rank():
+                    pre_compilation_list = tqdm(
+                        pre_compilation_list,
+                        desc="[Aiter Triton] Pre-compiling fp8 BMM kernel",
+                        total=max_batch_size,
+                    )
+
+                for m in pre_compilation_list:
```
Comment (ZhangLirong-amd) [inline atom/model_ops/attention_mla.py]: 1024 iteration calls torch.empty() twice here, it there any better method here? Can we use a buffer to slice?

File: atom/model_ops/attention_mla.py [hunk 6]
```diff
@@ -106,6 +110,7 @@ def dynamic_per_batched_tensor_quant(
     return x_scl_sat.to(dtype).contiguous(), scale.float().reciprocal()
 
 
+@MLASparseAttentionImplDecoratorForPluginMode
 @MLAAttentionImplDecoratorForPluginMode
 class MLAAttention(nn.Module):
```
Comment (zejunchen-zejun) [inline atom/model_ops/attention_mla.py:114]: The sparse MLA reuses the MLA interface right? Future we will reconstruct the decorator usage in ATOM OOT.

File: atom/plugin/attention_mla.py [hunk 7]
```diff
@@ -628,6 +628,19 @@ def forward_impl_plugin_mode(
         output=None,
     ):
         assert output is not None, "Output tensor must be provided."
+
+        # Dispatch to sparse MLA forward if this instance uses sparse attention
+        if getattr(self, "topk_indices_buffer", None) is not None:
```
Comment (zejunchen-zejun) [inline atom/plugin/attention_mla.py]: Could we find more clear way to identify if the instance is sparse MLA here?

File: atom/plugin/vllm/register.py (hunk 2 above)
Comment (kliuae-amd) [reply atom/plugin/vllm/register.py]: Done. Thanks for the suggestion.

File: atom/model_ops/attention_mla.py (hunk 4 above)
Comment (kliuae-amd) [reply atom/model_ops/attention_mla.py]: I think tqdm has already been used in a few places in ATOM's code base, but sure I can add it to the dependency list through this.

File: atom/model_ops/attention_mla.py (hunk 5 above)
Comment (kliuae-amd) [reply atom/model_ops/attention_mla.py]: Slicing from a buffer would result in different compiled caches from what non-sliced inputs expect at runtime, since for this kernel, tensor strides are passed to the kernel as int scalars, and triton can try to specialize them. Though this is performed at initialization and it does not contribute to runtime load.

File: atom/model_ops/attention_mla.py (hunk 6 above)
Comment (kliuae-amd) [reply atom/model_ops/attention_mla.py:114]: Yes it reuses the MLA interface to dispatch the forward method

File: atom/plugin/attention_mla.py (hunk 7 above)
Comment (kliuae-amd) [reply atom/plugin/attention_mla.py]: Sure I can use a flag marker instead to make the sparse MLA dispatch clearer

File: atom/model_ops/attentions/aiter_mla.py [hunk 8]
```diff
@@ -492,3 +498,74 @@ def build_for_cudagraph_capture(self, bs: int) -> AttentionMetaData:
             positions=positions, is_prefill=False, batch_size=bs, graph_bs=bs
         )
         return attn_matadata, context
+
+
+@AiterBackendDecoratorForPluginMode
+class AiterMLASparseBackend(AttentionBackend):
```
Comment (valarLip) [inline atom/model_ops/attentions/aiter_mla.py]: maybe move these stuff to atom/plugin/attention_mla.py

File: atom/model_ops/attention_mla.py (hunk 3 above)
Comment (valarLip) [reply atom/model_ops/attention_mla.py]: i don't like this part and i don't think we need these

**Review by valarLip — CHANGES_REQUESTED** [review]

File: atom/model_ops/attentions/aiter_mla.py (hunk 8 above)
Comment (kliuae-amd) [reply atom/model_ops/attentions/aiter_mla.py]: Done, moved these modules to attention backends for vllm plugin.

File: recipes/atom_vllm/GLM-5.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+vllm serve zai-org/GLM-5-FP8 \
+    --host localhost \
+    --port 8000 \
+    --tensor-parallel-size 8 \
+    --kv-cache-dtype fp8 \
+    --gpu_memory_utilization 0.9 \
+    --async-scheduling \
+    --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
+    --no-enable-prefix-caching \
+    --block-size 1
```
Comment (wuhuikx) [inline recipes/atom_vllm/GLM-5.md:27]: @ganyi1996ppo  could you please help comment, do we need block-size 1 here?

Comment (ganyi1996ppo) [reply recipes/atom_vllm/GLM-5.md:27]: oot seems only support `block-size 1`?

Comment (kliuae) [reply recipes/atom_vllm/GLM-5.md:27]: Yes currently only block_size 1 is supported for sparse mla oot

**Review by valarLip — APPROVED** [review]
